module/zfs/spa.c

```diff
@@ -6819,10 +6819,13 @@ spa_vdev_attach(spa_t *spa, uint64_t guid, nvlist_t *nvroot, int replacing,
 		return (spa_vdev_exit(spa, newrootvd, txg, error));
 
 	/*
-	 * Spares can't replace logs
+	 * log, dedup and special vdevs should not be replaced by spares.
 	 */
-	if (oldvd->vdev_top->vdev_islog && newvd->vdev_isspare)
+	if ((oldvd->vdev_top->vdev_alloc_bias == VDEV_BIAS_SPECIAL ||
+	    oldvd->vdev_top->vdev_alloc_bias == VDEV_BIAS_DEDUP ||
```

Comment:
I would do vdev_alloc_bias != VDEV_BIAS_NONE instead.  I just don't remember whether it also covers log for pools before special vdevs support or not.

Reply:
I checked and it does. I thought about doing the same but refrained from it if for some reason we want to extend vdev_alloc_bias. I updated it though as it indeed looks much cleaner now.